src/rulesets/Base/mapreduce.jl

```diff
@@ -67,30 +67,55 @@ function rrule(
 end
 
 function rrule(
-    config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f, xs::AbstractArray; dims=:
-)
-    fx_and_pullbacks = map(x->rrule_via_ad(config, f, x), xs)
-    y = sum(first, fx_and_pullbacks; dims=dims)
+    config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f::F, xs::AbstractArray; dims=:
+) where {F}
+    project = ProjectTo(xs)
 
-    pullbacks = last.(fx_and_pullbacks)
+    if _uses_input_only(f, eltype(xs))
+        # Then we can compute the forward pass as usual, save nothing but `xs`:
+        y = sum(f, xs; dims=dims)
+        function sum_pullback_easy(dy)
+            dxs = unthunk(dy) .* conj.(only.(only.(derivatives_given_output.(nothing, f, xs))))
+            return (NoTangent(), NoTangent(), project(dxs))
+        end
+        return y, sum_pullback_easy
+    end
 
-    project = ProjectTo(xs)
+    # In the general case, we need to save all the pullbacks:
+    fx_and_pullbacks = map(x -> rrule_via_ad(config, f, x), xs)
+    y = sum(first, fx_and_pullbacks; dims=dims)
+
+    function sum_pullback_f(dy)
+        # For arrays of arrays, we ought to protect the element against broadcasting:
+        dys = dims isa Colon ? Ref(unthunk(dy)) : unthunk(dy)
+        if Base.issingletontype(F)
+            # Then at least `f` has no gradient. Note that broadcasting here
+            # gets the shape right with or without `dims` keyword.
+            dxs = broadcast(fx_and_pullbacks, dys) do (_, back), dy1
+                unthunk(last(back(dy1)))
+            end
+            return (NoTangent(), NoTangent(), project(dxs))
 
-    function sum_pullback(ȳ)
-        call(f, x) = f(x)
-        # if dims is :, then need only left-handed only broadcast
-        broadcast_ȳ = dims isa Colon  ? (ȳ,) : ȳ
-        f̄_and_x̄s = call.(pullbacks, broadcast_ȳ)
-        # no point thunking as most of work is in f̄_and_x̄s which we need to compute for both
-        f̄ = if fieldcount(typeof(f)) === 0 # Then don't need to worry about derivative wrt f
-            NoTangent()
         else
-            sum(first, f̄_and_x̄s)
+            # Most general case. If `f` were stateful, we would need to reverse the order
+            # of iteration here, but since this funciton makes no guarantee, even the primal
+            # result is then ill-defined.
+            df_and_dxs = broadcast(fx_and_pullbacks, dys) do (_, back), dy1
+                map(unthunk, back(dy1))
+            end
+            df = sum(first, df_and_dxs)
+            dxs = map(last, df_and_dxs)
+            return (NoTangent(), df, project(dxs))
         end
-        x̄s = map(unthunk ∘ last, f̄_and_x̄s) # project does not support receiving InplaceableThunks
-        return NoTangent(), f̄, project(x̄s)
     end
-    return y, sum_pullback
+    return y, sum_pullback_f
+end
+
+function _uses_input_only(f::F, ::Type{xT}) where {F,xT}
```

Comment:
what is the name of the class of functions that this is true for?
Is it just linear operators?
Reversable functions?

Reply:
This will be impossible for some functions, like `sin`, and I guess any continuous function which isn't one-to-one. 

But it's often a property of implementations, e.g. for `y = tanh(x)` you can very cheaply get the gradient from `y`, so the rule does, instead of re-computing, and hence this gives `false`.

Reply:
Could we add a docstring to explain what this function does? I suspect it "returns `true` if the gradient of `f` can be [1] computed by only considering the inputs of type `xT`", but I suspect this definition is missing the word "efficiently" at [1]?

It seems to be really looking at the implementation of `derivatives_given_output` of `F`. It also assumes that the `derivatives_given_output` is type stable, right? Otherwise it returns `false` since we get an `Union`?

e.g.
```julia
julia> myfunc(val) = ifelse(val > 0, 1.0, 1)
myfunc (generic function with 1 method)

julia> Core.Compiler._return_type(myfunc, Tuple{Float64})
Union{Float64, Int64}
```

As far as I understand in this case the final answer isn't wrong, it's just not computed optimally?

Reply:
```suggestion
"""
    _uses_input_only(f, xT::Type)

Returns `true` if it can prove that `derivatives_given_output` will work using only the input
of the given type. Thus there is no need to store the output `y = f(x::xT)`, allowing us to take
a fast path in the `rrule` for `sum(f, xs)`.

Works by seeing if the result of `derivatives_given_output(nothing, f, x)` can be inferred.
The method of `derivatives_given_output` usually comes from `@scalar_rule`.
"""
function _uses_input_only(f::F, ::Type{xT}) where {F,xT}
```